### PR TITLE
libvirt: 5.4.0 -> 6.1.0

### LINKED ISSUE #81102: libvirt is too old for qemu in unstable
Currently, qemu version is 4.2.0 and libvirt is 5.4.0. As I've already mentioned [here](https://github.com/NixOS/nixpkgs/pull/76065#issuecomment-569632713) there are some incompatibilities between qemu >= 4.1 and libvirt < 5.6.0. In some cases it may significantly affect performace (e.g. typical windows vm with vfio).

The latest libvirt version is [6.0.0](https://libvirt.org/news.html#v6.0.0). So, I think it's time to upgrade.

Unfortunately, it seems that something has changed in libvirt build process and just changing version/hash in existing derivation is not enough.

/cc @fpletz @globin (libvirt maintainers)

### PR DESCRIPTION
**Pull request by wedens**
###### Motivation for this change

Closes #81102

###### Things done

I'm unable to test on Darwin, but I tested with `buildFromTarball` set to true and false.

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

/cc @fpletz @globin

### CODE REVIEW

**Comment by wedens** [comment]
Hopefully it can make it into 20.03 to avoid qemu co-existing with old libvirt, but I guess there is just not enough time to extensively test this update.

**Comment by Izorkin** [comment]
@wedens fixed this bug? - https://github.com/NixOS/nixpkgs/pull/66584#issuecomment-524656760

**Comment by wedens** [comment]
@Izorkin Unfortunately, no idea. I don't use (and can't easily setup) bridge networking to test it. If you (or someone else who uses bridge) can test, that'd be neat.

**Comment by Izorkin** [comment]
@wedens error build  pythonPackages.libvirt
```
Executing setuptoolsBuildPhase
libvirt-python requires Python 3.x to build
```

**Comment by Izorkin** [comment]
If rebuild systev used python2Packages.libvirt, need change to python3Packages.libvirt

**Comment by jonringer** [comment]
please freeze the 5.4.0 version for python 2

```
[41 built (1 failed), 271 copied (5085.6 MiB), 916.4 MiB DL]
error: build of '/nix/store/adv1j8r1342895ysgviibjbbnl5y8d4b-env.drv' failed
https://github.com/NixOS/nixpkgs/pull/81161
4 package failed to build:
nixops nixopsUnstable nixops_1_6_1 python27Packages.libvirt

33 package built:
collectd collectd-data docker-machine-kvm docker-machine-kvm2 gnome3.gnome-boxes haskellPackages.libvirt-hs libguestfs libguestfs-with-appliance libvirt libvirt-glib libvmi minikube minishift perl528Packages.SysVirt perl530Packages.SysVirt python27Packages.guestfs python37Packages.guestfs python37Packages.libvirt python38Packages.guestfs python38Packages.libvirt rubyPackages.ruby-libvirt rubyPackages_2_5.ruby-libvirt rubyPackages_2_7.ruby-libvirt terraform-full terraform-providers.libvirt terraform_0_11-full terragrunt vagrant virt-manager virt-manager-qt virt-top virt-viewer virtlyst
```

**Comment by Izorkin** [comment]
@ With 5.9.0 correct builded

**Comment by wedens** [comment]
Yeah, python2 support was dropped in 6.0.0. I'll try to figure out how to use different versions for python2/3. Hopefully it won't make it confusing.

**Comment by wedens** [comment]
Alright, I've added 5.9.0 for python2. Not sure if it's the correct thing to do though (how possible that there will be some incompatibilities between different libvirt and libvirt-python versions?).

**Comment by Izorkin** [comment]
@wedens error rebuild
```error: assertion failed at ../pkgs/development/python-modules/libvirt/5.9.0.nix:8:9```

**Comment by Izorkin** [comment]
@GrahamcOfBorg build nixops

**Comment by Izorkin** [comment]
@wedens error build. Used libvirt 6.0.0, needed 5.9.0
```
Executing setuptoolsBuildPhase
running bdist_wheel
running build
/nix/store/w6pfcj87889hgz0f77blq59z85mdrsh8-pkg-config-0.29.2/bin/pkg-config --print-errors --atleast-version=0.9.11 libvirt
/nix/store/bh447min7f6vqdw3x1iywr1vxyx4ix8y-python-2.7.17/bin/python2.7 generator.py libvirt /nix/store/4aqz37c2fdg3klqhxhji707ypzdcra96-libvirt-6.0.0/share/libvirt/api/libvirt-ap$
Found 478 functions in /nix/store/4aqz37c2fdg3klqhxhji707ypzdcra96-libvirt-6.0.0/share/libvirt/api/libvirt-api.xml
Found 0 functions in libvirt-override-api.xml
Generated 389 wrapper functions
/nix/store/bh447min7f6vqdw3x1iywr1vxyx4ix8y-python-2.7.17/bin/python2.7 generator.py libvirt-qemu /nix/store/4aqz37c2fdg3klqhxhji707ypzdcra96-libvirt-6.0.0/share/libvirt/api/libvi$
Found 5 functions in /nix/store/4aqz37c2fdg3klqhxhji707ypzdcra96-libvirt-6.0.0/share/libvirt/api/libvirt-qemu-api.xml
Found 0 functions in libvirt-qemu-override-api.xml
Generated 3 wrapper functions
Traceback (most recent call last):
  File "generator.py", line 2143, in <module>
    qemuBuildWrappers(sys.argv[1])
  File "generator.py", line 2008, in qemuBuildWrappers
    items.sort(key=lambda i: (int(i[1]), i[0]))
  File "generator.py", line 2008, in <lambda>
    items.sort(key=lambda i: (int(i[1]), i[0]))
ValueError: invalid literal for int() with base 10: 'VIR_DOMAIN_AGENT_RESPONSE_TIMEOUT_BLOCK'
error: command '/nix/store/bh447min7f6vqdw3x1iywr1vxyx4ix8y-python-2.7.17/bin/python2.7' failed with exit status 1

```

**Comment by wedens** [comment]
I'm not sure what to do about it, tbh.

I'd rather not use a band aid like updating to 5.9.0 instead of  6.0.0 and putting of the python2 problem for later.

**Comment by jonringer** [comment]
> I'd rather not use a band aid like updating to 5.9.0 instead of 6.0.0 and put off the python2 problem for later.

If it was just the python-module, i would be fine with deprecating. But I'm not okay with breaking nixops packages

**Comment by wedens** [comment]
What are chances that nixops will update to python 3 in the nearby future?

I guess it'll be reasonable compromise to update to 5.9.0 for now and to 6.* closer to 20.09 release.

**Comment by jonringer** [comment]
Since the python2 was not the goal of this PR, you could also just do a freeze at 5.4.0, instead of bumping to 5.9.0

**Comment by wedens** [comment]
@jonringer Are you suggesting having two libvirt/libvirt-python versions (5.4.0 and 6.0.0) in nixpkgs?

I'm not sure how this compatibility works, but as far as I understand, if libvirt daemon (presumably used via libvirt nixos module) is 6.0.0, nixops (that'll use libvirt-python with libvirt 5.4.0), won't be able to manage this libvirt daemon (which is 6.0.0).

**Comment by jonringer** [comment]
I suggesting that, instead of freezing the python2 version at 5.9.0, you keep it at 5.4.0. python3 will be at latest

**Comment by wedens** [comment]
Ok. I've added `libvirt_5_4_0` which is used as a dependency for python-libvirt with python2.

**Comment by jonringer** [comment]
> @jonringer Are you suggesting having two libvirt/libvirt-python versions (5.4.0 and 6.0.0) in nixpkgs?
> 
> I'm not sure how this compatibility works, but as far as I understand, if libvirt daemon (presumably used via libvirt nixos module) is 6.0.0, nixops (that'll use libvirt-python with libvirt 5.4.0), won't be able to manage this libvirt daemon (which is 6.0.0).

Just read this. Hmm, we should get someone from nixops to chime in. Forgot that libvirt was more than just a library. 

cc @grahamc @edolstra

**Comment by jonringer** [comment]
It does however, fix all applications.

```
[14 built, 2 copied (56.6 MiB), 12.0 MiB DL]
https://github.com/NixOS/nixpkgs/pull/81161
38 package built:
collectd collectd-data docker-machine-kvm docker-machine-kvm2 gnome3.gnome-boxes haskellPackages.libvirt-hs libguestfs libguestfs-with-appliance libvirt libvirt-glib libvirt_5_4_0 libvmi minikube minishift nixops nixopsUnstable nixops_1_6_1 perl528Packages.SysVirt perl530Packages.SysVirt python27Packages.guestfs python27Packages.libvirt python37Packages.guestfs python37Packages.libvirt python38Packages.guestfs python38Packages.libvirt rubyPackages.ruby-libvirt rubyPackages_2_5.ruby-libvirt rubyPackages_2_7.ruby-libvirt terraform-full terraform-providers.libvirt terraform_0_11-full terragrunt vagrant virt-manager virt-manager-qt virt-top virt-viewer virtlyst
```

**Comment by volth** [comment]
`6.1.0` released

**Comment by Izorkin** [comment]
With 6.0.0 frezing start server.

**Comment by volth** [comment]
> With 6.0.0 frezing start server.

Is it the same issue we had trying to bump `5.4.0`->`5.6.0` (https://github.com/NixOS/nixpkgs/pull/66584) ?
You offered a fix then, does it help with `6.0.0` ?

**Comment by wedens** [comment]
> With 6.0.0 frezing start server.

Any details?

**Comment by volth** [comment]
> I suggesting that, instead of freezing the python2 version at 5.9.0, you keep it at 5.4.0. python3 will be at latest

If we freeze `libvirt-5.4.0`, should we freeze `qemu-4.1` for it (per https://github.com/NixOS/nixpkgs/issues/81102) ?

**Comment by Izorkin** [comment]
> > With 6.0.0 frezing start server.
> 
> Is it the same issue we had trying to bump `5.4.0`->`5.6.0` (#66584) ?
> You offered a fix then, does it help with `6.0.0` ?

Yes, this error.
My fix not full worked. After `nixos-rebuild` swith need manual runnig `systemctl reload br0-netdev`

**Comment by volth** [comment]
Probably the easiest solution would be to find the commit between `5.4.0` and `5.6.0` which introduced the problem and revert it

**Comment by Izorkin** [comment]
@volth new version libvirt used systemd socket activation - https://bugzilla.redhat.com/show_bug.cgi?id=1750340

**Comment by volth** [comment]
> new version libvirt used systemd socket activation -
```
Thus if the mgmt app / admin wants to use TCP/TLS sockets they have two choices

  - To continue the old approach (setting --listen in /etc/sysconfig/libvirtd), then they MUST use 'systemctl mask ...' for all the socket units listed above, before libvirtd.service is started.

  - To adapt to use the new approach, then don't touch /etc/sysconfig/libvirtd at all. Instead use 'systemctl enable libvirtd-tls.socket && 'systemctl start libvirtd-tls.socket'   (assuming  TLS is desired).
```

I'd prefer the first option (keep the daemon running) for seamless upgrade.
Eventually we could switch to systemd socket activation

**Comment by volth** [comment]
There is not only socket activation, there is also `virtqemud` daemon we have to start

**Comment by volth** [comment]
https://github.com/wedens/nixpkgs/pull/1

libvirtd rejects to start with --listen flag in presence of systemd, so I switched to socket activation.
it works on my laptop, but I have not tested yet various cases like restoring bridges after nixos-rebuild

**Comment by volth** [comment]
`./bootstrap` script is gone in 6.1.0.
it was the script which failed on Darwin and that's why a tarball (with post-`./bootstrap` state) was required.
There is a good chance that Darwin build of libvirt-6.1.0 will succeed from git sources.
Then tarball can be removed.

**Comment by Izorkin** [comment]
With libvirt 5.9.0 and 6.1.0 correct work.

**Comment by volth** [comment]
also, `collectd` needs to be upgraded.
its `virt` plugin logs protocol errors when interacting with libvirt-6.1.0

**Comment by volth** [comment]
https://github.com/wedens/nixpkgs/pull/2 cosmetic

**Comment by mmilata** [comment]
Thank you for the effort to bring recent libvirt into 20.03! I've briefly tested this branch and things seem to work fine - installing QEMU/KVM virtual machine and using it with the default NATted network as well as NixOS-manged bridge.

**Comment by volth** [comment]
I;d be happy to get rid of systemd socket activation eventually.
It makes `virsh` unreliable, sometimes it get stuck :(

**Comment by jonringer** [comment]
Personally, i would like to see the merge commits be removed, should be able to clean up with `git rebase -i`.

otherwise LGTM
```
[20 built, 0.0 MiB DL]
https://github.com/NixOS/nixpkgs/pull/81161
38 package built:
collectd collectd-data docker-machine-kvm docker-machine-kvm2 gnome3.gnome-boxes haskellPackages.libvirt-hs libguestfs libguestfs-with-appliance libvirt libvirt-glib libvirt_5_4_0 libvmi minikube minishift nixops nixopsUnstable nixops_1_6_1 perl528Packages.SysVirt perl530Packages.SysVirt python27Packages.guestfs python27Packages.libvirt python37Packages.guestfs python37Packages.libvirt python38Packages.guestfs python38Packages.libvirt rubyPackages.ruby-libvirt rubyPackages_2_5.ruby-libvirt rubyPackages_2_7.ruby-libvirt terraform-full terraform-providers.libvirt terraform_0_11-full terragrunt vagrant virt-manager virt-manager-qt virt-top virt-viewer virtlyst
```

**Comment by wedens** [comment]
> Personally, i would like to see the merge commits be removed, should be able to clean up with git rebase -i.

Done.

**Comment by wedens** [comment]
> It makes virsh unreliable, sometimes it get stuck :(

@volth Is it some known issue with libvirt itself?

**Comment by volth** [comment]
> > It makes virsh unreliable, sometimes it get stuck :(
> 
> @volth Is it some known issue with libvirt itself?

No, it is my `5.4.0`->`6.1.0` migration experience.

**Comment by Izorkin** [comment]
Really upgrade 5.4 to 5.9 and 6.1?

**Comment by volth** [comment]
> upgrade 5.4 to 5.9 and 6.1?

I think we have no.
After this PR, NixOS module won't work with 5.4 because it does not start `libvirtd.service`.
It relies on systemd socket activation.

**Comment by wedens** [comment]
Alright, let's do it while we're at it. @Izorkin I've applied your patch

File: nixos/modules/virtualisation/libvirtd.nix [hunk 1]
```diff
@@ -214,14 +214,14 @@ in {
...
     systemd.services.libvirtd = {
-      description = "Libvirt Virtual Machine Management Daemon";
-
-      wantedBy = [ "multi-user.target" ];
       requires = [ "libvirtd-config.service" ];
       after = [ "systemd-udev-settle.service" "libvirtd-config.service" ]
               ++ optional vswitch.enable "ovs-vswitchd.service";
 
-      environment.LIBVIRTD_ARGS = ''--config "${configFile}" ${concatStringsSep " " cfg.extraOptions}'';
+      environment.LIBVIRTD_ARGS = concatMapStringsSep " " escapeShellArg (
```
Comment (conferno) [inline nixos/modules/virtualisation/libvirtd.nix]:
```suggestion
      environment.LIBVIRTD_ARGS = escapeShellArgs (
```

**Comment by wedens** [comment]
How can we move this PR forward?

**Comment by nh2** [comment]
What's remaining here?

Would be very nice to have working libvirt in 20.03!

**Comment by globin** [comment]
I'll work on this first thing on monday

**Comment by worldofpeace** [comment]
@globin Perfect, I'd like to see this in 20.03 so we can have this working. Thanks.

**Comment by volth** [comment]
I am using this in prod for 2 weeks.
Some hosts with old VMs required 2-3 reboots to get into a stable state.
Fresh installs work well.

**Review by FRidh — APPROVED** [review]
Python bit looks good. I haven't checked the rest.

**Pull request merged**